builder/builder.go

```diff
@@ -290,71 +291,107 @@ func (b *Builder) parseWithdrawalMessages(
+				err := b.storeWithdrawalMsgInEVM(execTxResult, ethState, header)
 				if err != nil {
 					return nil, fmt.Errorf("store withdrawal msg in EVM: %v", err)
 				}
-
-				// Populate the nonce in the tx event attributes.
-				for i := range execTxResult.Events {
-					event := &execTxResult.Events[i] // Get a pointer to the event, so we can modify it.
-					if event.Type == rolluptypes.EventTypeWithdrawalInitiated {
-						event.Attributes = append(event.Attributes, abcitypes.EventAttribute{
-							Key:   rolluptypes.AttributeKeyNonce,
-							Value: hexutil.Encode(nonce.Bytes()),
-						})
-					}
-				}
 			}
 		}
 	}
 	return execTxResult, nil
 }
 
-// storeWithdrawalMsgInEVM stores the withdrawal message hash in the monomer evm state db and returns the L2ToL1MessagePasser
-// message nonce used for the withdrawal. This is used for proving withdrawals.
+// storeWithdrawalMsgInEVM stores the withdrawal message hash in the monomer evm state db and appends the L2ToL1MessagePasser
+// message nonce used for the withdrawal to the tx events. This is used for proving withdrawals.
 func (b *Builder) storeWithdrawalMsgInEVM(
-	withdrawalMsg *rolluptypes.MsgInitiateWithdrawal,
+	execTxResult *abcitypes.ExecTxResult,
 	ethState *state.StateDB,
 	header *monomer.Header,
-) (*big.Int, error) {
+) error {
 	monomerEVM, err := evm.NewEVM(ethState, header)
 	if err != nil {
-		return nil, fmt.Errorf("new EVM: %v", err)
+		return fmt.Errorf("new EVM: %v", err)
 	}
 	executer, err := bindings.NewL2ToL1MessagePasserExecuter(monomerEVM)
 	if err != nil {
-		return nil, fmt.Errorf("new L2ToL1MessagePasserExecuter: %v", err)
+		return fmt.Errorf("new L2ToL1MessagePasserExecuter: %v", err)
 	}
 
 	// Get the current message nonce before initiating the withdrawal.
 	messageNonce, err := executer.GetMessageNonce()
 	if err != nil {
-		return nil, fmt.Errorf("get message nonce: %v", err)
+		return fmt.Errorf("get message nonce: %v", err)
 	}
 
-	senderCosmosAddress, err := sdk.AccAddressFromBech32(withdrawalMsg.GetSender())
+	var event *abcitypes.Event
+	for i := range execTxResult.Events {
+		event = &execTxResult.Events[i]
+		if event.Type == rolluptypes.EventTypeWithdrawalInitiated {
+			break
+		}
+	}
+	if event == nil {
+		return fmt.Errorf("withdrawal event not found")
+	}
+
+	params, err := parseWithdrawalEventAttributes(event)
 	if err != nil {
-		return nil, fmt.Errorf("convert sender to cosmos address: %v", err)
+		return fmt.Errorf("parse withdrawal attributes: %v", err)
 	}
 
 	// Initiate the withdrawal in the Monomer ethereum state.
-	if err = executer.InitiateWithdrawal(
-		common.BytesToAddress(senderCosmosAddress.Bytes()),
-		withdrawalMsg.Value.BigInt(),
-		common.HexToAddress(withdrawalMsg.GetTarget()),
-		new(big.Int).SetBytes(withdrawalMsg.GasLimit),
-		withdrawalMsg.GetData(),
-	); err != nil {
-		return nil, fmt.Errorf("initiate withdrawal: %v", err)
+	if err = executer.InitiateWithdrawal(*params); err != nil {
```

Comment:
Let's make `InitiateWithdrawal` accept a pointer to avoid copying (maybe the compiler is smart and I'm missing something, but we use pointers elsewhere anyway).

Reply:
Yeah good call, I'll update this